manuals/esedbinfo.1
.Dd June 25, 2026
.Dt ESEDBINFO 1
.Os
.Sh NAME
.Nm esedbinfo
.Nd determines information about an Extensible Storage Engine (ESE) Database \
(EDB) file
.Sh SYNOPSIS
.Nm esedbinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm esedbinfo
is a utility to determine information about an Extensible Storage Engine (ESE) \
Database (EDB) file
.Pp
.Nm esedbinfo
is part of the
.Nm libesedb
package.
.Nm libesedb
is a library to access the Extensible Storage Engine (ESE) Database File (EDB) \
format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# esedbinfo test.edb
esedbinfo 20260625
.sp
Extensible Storage Engine Database information:
        File type:              Database
        Created in format:      0x620,12
        Current format:         0x620,12
        Page size:              8192 bytes
.sp
Catalog content:
        Number of tables:       1
.sp
Table: 1                        MSysObjects (2)
        Amount of columns:      22
        Column: 1               ObjidTable (1)
        Column: 2               Type (2)
        Column: 3               Id (3)
        Column: 4               ColtypOrPgnoFDP (4)
        Column: 5               SpaceUsage (5)
        Column: 6               Flags (6)
        Column: 7               PagesOrLocale (7)
        Column: 8               RootFlag (8)
        Column: 9               RecordOffset (9)
        Column: 10              LCMapFlags (10)
        Column: 11              KeyMost (11)
        Column: 12              Name (128)
        Column: 13              Stats (129)
        Column: 14              TemplateTable (130)
        Column: 15              DefaultValue (131)
        Column: 16              KeyFldIDs (132)
        Column: 17              VarSegMac (133)
        Column: 18              ConditionalColumns (134)
        Column: 19              TupleLimits (135)
        Column: 20              Version (136)
        Column: 21              CallbackData (256)
        Column: 22              CallbackDependencies (257)
.sp
        Amount of indexes:      3
        Index: 1                Id (2)
        Index: 2                Name (4)
        Index: 3                RootObjects (5)
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr esedbexport 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libesedb/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add LZXPRESSED compressed string to compression test
* add leaf value (data definition) cache?
* use leaf page previous/next values?

* fix XOR checksum in esedb_test_checksum_calculate_little_endian_ecc32 test
* libfdata next search keep going on a key match?

* fix flaky test for System_Search_AutoSummary (one byte off?)
* pyesedb
  - null value support
  - long value: add data size support
* refactor libesedb_record_get_value_binary_data_size and
  libesedb_record_get_value_binary_data
* complete libfdata btree refactor
  - libesedb_values_tree.c replace by fdata functions?
  - fix is_flexible_match for xp_search.edb indexes
* record flags
  - add indication of corrupt records

* format notes:
  - value of un-initialized page checksum: 0xdeadbeefbaadf00d
    "ESE Super ECCXORChecksum"

* create tool to analyze cache usage, hits and misses
* move read of data definition into leaf value callback
  - create separate callback for long values?
  - this will clean up data definition
* refactor:
  - libesedb_multi_value_get_value what about value_type ?
  - clean up multi value calls to libfvalue
  - libesedb_long_value_get_segment_data
* deprecate:
  - libesedb_record_get_value
  - libesedb_multi_value_get_value
* complete libfvalue update integration - multi value

* windows search add support for windows 8 SystemIndex_PropertyStore
* libesedb check esedbtools_libX header files for _LIBESEDB_LIBX_ defs
* handle compressed values inside the library
  - compressed binary data
* esedbexport -t /root creates /root.export ?
* esedbexport skip table on error
* check Win2K ntds.dit support
* clean up and refactor info handle
* export handle: fix creation of indexes directory

API:
* add function to retrieve database signature
  - can be used to match up EDB/STM pairs
* add long value functions (string and binary data?)
* remove
  - libesedb_record_get_value
  - libesedb_long_value_get_segment_data
  - libesedb_multi_value_get_value

* windows search
  - multi long value segment decryption/compression support
* export error in more test files: nibble count value exceeds maximum

* code clean up
  - change page value index and number of values into int
  - clean up TODO's
  - refactor info handle
* combine libesedb_io_handle_read_page with libesedb_page_read
  - need libesedb_page_read for recovery ?

* page tree on demand
  - check sanity of page value 0 
* clean up
  - libesedb_data_type_definition remove offset
* handle default value
  - need database which uses default value
* parse variable types even when they are not part of the table catalog definition
* add support for build in table number 1
* implement cloning of items for improved asynchronous and multi threaded usage

libesedb:
* add get_table_by_object_identifier function
* ECC checksum
  - non 4 byte modulus buffer support
  - add alignment optimization
  - error correction
  - Windows 7 extended checksum ?
* template table
  - what about non-tagged data types
* multi value
  - what about multi value 0x18
* database header
  - compare primary and secondary
* catalog
  - compare primary and secondary
  - limit the size of the catalog ?
  - support for type 0x0006
  - support for type 0x0007 (line 3457 in libesedb_page_tree.c)
* long values
  - support for non-linear stored long value segments - current solution does a sanity check
* default values
   - apply to data definition

Error tolerance
* probe test for the page size using the backup database header
* use backup catalog

Debug
* is function libesedb_debug_print_table_group_of_bits needed?
* print value flags
* print part of keys as strings

Tools
* esedbinfo
  - page summary (scan for uninitialized, empty, etc. pages)
  - file header summary (check backup file header)
  - summary of catalog (check backup catalog)
  - print database state
* esedbexport
  - add export mode

Windows search:
* improve windows_search_decompress_byte_indexed_compressed_data

File format
* where is the root object defined? or is it implied to be object 1?
* compressed table, column, index names ?
* page flags
  - 0x0400, 0x0800, 0x8000
* tagged values
  - flag 0x4000
* multi value offsets
  - flag 0x8000
* page flags
* page keys
* (revision 0x11) extended page header
* what about the first byte in 7-bit ascii compressed data
* codepage 1200 string in Exchange EDB starts with 0xff 0xff
* streaming file
  - only consists of ESEDB header with paged streamed data,
    how to reconstruct streams ?

Deployment:
* update configure
  - improve libfdata detection
  - improve libfguid detection
  - improve libfwnt detection
* update manuals
  - add esedbexport.1

Performance
* create data definition tree in page tree for faster lookup of keys
  keep list for sequential access (B+-tree)
* create a UTF-8 version of run-length compressed UTF-16 string decompression function
* use register alignment for key comparison
* optimize libesedb_checksum_calculate_little_endian_ecc32

Test suite:
* windows search: work on index-based compression
  - create test program to validate output

Notes
* create a root page table instead?
  - no
* read catalog data type definitions and convert them into column definitions?
  - no

20090913
* initial version based on liblnk
* see `git log' for more recent change log



esedbtools/esedbinfo.c
/*
 * Shows information obtained from an Extensible Storage Engine (ESE) Database (EDB) file.
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "esedbtools_getopt.h"
#include "esedbtools_libcerror.h"
#include "esedbtools_libclocale.h"
#include "esedbtools_libcnotify.h"
#include "esedbtools_libesedb.h"
#include "esedbtools_output.h"
#include "esedbtools_signal.h"
#include "esedbtools_unused.h"
#include "info_handle.h"

info_handle_t *esedbinfo_info_handle = NULL;
int esedbinfo_abort                  = 0;

/* Signal handler for esedbinfo
 */
void esedbinfo_signal_handler(
      esedbtools_signal_t signal ESEDBTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "esedbinfo_signal_handler";

	ESEDBTOOLS_UNREFERENCED_PARAMETER( signal )

	esedbinfo_abort = 1;

	if( esedbinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     esedbinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description    = \
		"Use esedbinfo to determine information about an Extensible Storage Engine (ESE) Database (EDB) file.";

	esedbtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libesedb_error_t *error    = NULL;
	system_character_t *source = NULL;
	char *program              = "esedbinfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( esedbtools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "esedbtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( esedbtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	esedbtools_output_version_fprint(
	 stdout,
	 program );

	if( esedbtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = esedbtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				esedbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				esedbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				esedbtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		esedbtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libesedb_notify_set_stream(
	 stderr,
	 NULL );
	libesedb_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &esedbinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open(
	     esedbinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     esedbinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     esedbinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &esedbinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( esedbinfo_info_handle != NULL )
	{
		info_handle_free(
		 &esedbinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

